src/base/async/int.c

```diff
@@ -1929,100 +1927,25 @@ static void redirect_devices(void)
 
 static int redir_it(void)
 {
-    uint16_t lol_lo, lol_hi, sda_lo, sda_hi, sda_size, redver, mosver;
-    uint8_t major, minor;
-    int is_MOS;
-    int is_cf;
-
     if (redir_state)
 	return 0;
     redir_state++;
 
-    /*
-     * To start up the redirector we need
-     * (1) the list of list,
-     * (2) the DOS version and
-     * (3) the swappable data area.
-     */
     pre_msdos();
 
-    LWORD(eax) = 0x1100;
-    do_int_call_back(0x2f);
-    if (LO(ax) != 0xff) {
-	error("Redirector unavailable\n");
-	post_msdos();
-	return 0;
-    }
-
-    LWORD(eax) = 0x5200;
-    call_msdos();
-    ds_printf
-	("INT21 +1 at %04x:%04x: AX=%04x, BX=%04x, CX=%04x, DX=%04x, DS=%04x, ES=%04x\n",
-	 SREG(cs), LWORD(eip), LWORD(eax), LWORD(ebx),
-	 LWORD(ecx), LWORD(edx), SREG(ds), SREG(es));
-    lol_lo = LWORD(ebx);
-    lol_hi = SREG(es);
-
-    LWORD(eax) = LWORD(ebx) = LWORD(ecx) = LWORD(edx) = 0x3099; // MOS version trigger
-    call_msdos();
-    ds_printf
-	("INT21 +2b at %04x:%04x: AX=%04x, BX=%04x, CX=%04x, DX=%04x, DS=%04x, ES=%04x\n",
-	 SREG(cs), LWORD(eip), LWORD(eax), LWORD(ebx),
-	 LWORD(ecx), LWORD(edx), SREG(ds), SREG(es));
-    mosver = LWORD(eax);
-    LWORD(eax) = 0x3000;                                        // DOS version std request
-    call_msdos();
-    ds_printf
-	("INT21 +2 at %04x:%04x: AX=%04x, BX=%04x, CX=%04x, DX=%04x, DS=%04x, ES=%04x\n",
-	  SREG(cs), LWORD(eip), LWORD(eax), LWORD(ebx),
-	 LWORD(ecx), LWORD(edx), SREG(ds), SREG(es));
-    major = LO(ax);
-    minor = HI(ax);
-    is_MOS = (LWORD(eax) != mosver); // different result!
+    LWORD(eax) = DOS_HELPER_MFS_HELPER;
+    LWORD(ebx) = DOS_SUBHELPER_MFS_REDIR_INIT;
+    mfs_inte6();
 
-    LWORD(eax) = 0x5d06;
-    call_msdos();
-    ds_printf
-	("INT21 +3 at %04x:%04x: AX=%04x, BX=%04x, CX=%04x, DX=%04x, DS=%04x, ES=%04x\n",
-	 SREG(cs), LWORD(eip), LWORD(eax), LWORD(ebx),
-	 LWORD(ecx), LWORD(edx), SREG(ds), SREG(es));
-    sda_lo = LWORD(esi);
-    sda_hi = SREG(ds);
-    sda_size = LWORD(ecx);
-
-    ds_printf("INT21: lol = 0x%04x\n", (lol_hi << 4) + lol_lo);
-    ds_printf("INT21: sda = 0x%04x, size = 0x%04x\n",
-	      (sda_hi << 4) + sda_lo, sda_size);
-    ds_printf("INT21: ver = 0x%02x, 0x%02x\n", major, minor);
-
-    /* Figure out the redirector version */
-    if (is_MOS) {
-        redver = REDVER_NONE;
+    if (!isset_CF()) {	/* We have a functioning redirector so use it */
+      LWORD(eax) = DOS_HELPER_MFS_HELPER;
+      LWORD(ebx) = DOS_SUBHELPER_MFS_REDIR_CLONE;
+      mfs_inte6();
```

Comment:
`do_intr_call_back()` would be better.

Reply:
I don't seem to be able to swap this, with 
```C
diff --git a/src/base/async/int.c b/src/base/async/int.c
index 80f35e50a..62dac58b3 100644
--- a/src/base/async/int.c
+++ b/src/base/async/int.c
@@ -1942,12 +1942,12 @@ static int redir_it(void)
 
     LWORD(eax) = DOS_HELPER_MFS_HELPER;
     LWORD(ebx) = DOS_SUBHELPER_MFS_INIT;
-    mfs_inte6();
+    do_int_call_back(DOS_HELPER_INT);
 
     if (!isset_CF()) { /* We have a functioning redirector so use it */
       LWORD(eax) = DOS_HELPER_MFS_HELPER;
       LWORD(ebx) = DOS_SUBHELPER_MFS_REDIRECT_DEVICES;
-      mfs_inte6();
+      do_int_call_back(DOS_HELPER_INT);
     } else {
       ds_printf("INT21: this DOS has an incompatible redirector\n");
     }
```

I get 
~~~
dosemu.bin: do_vm86.c:611: callback_return: Assertion `callback_level > 0' failed.
~~~

Any idea what I'm doing wrong?

Reply:
No, don't know what is it: I now
committed such change to git and
nothing broke. Maybe in your case
things like `pre_msdos()` got called
recursively, while they store only one
copy of context.

Reply:
Yes, that's exactly it. Switching to `do_int_call_back()` meant that calls to `pre/post_msdos()` were not recursive, but nested once.

Reply:
I pushed a new version that avoids the issue for me.
instead of asserting I wonder if it would be too much trouble to have pre/post_msdos() take a regs parameter? But I suppose if that were done, we may as just do saved_regs=REGS and REGS=saved_regs.